fix: change setup for read only tasks

jhonasiv opened the pull request:
Instead of always considering that a read_only task setup is done, we will consider that it is done whenever its not-read only counterpart is running. Also, read_only tasks do nothing on perform_setup.

jhonasiv committed:
fix: change setup for read only tasks

Instead of always considering that a read_only task setup is done, we
will consider that it is done whenever its not-read only counterpart is
running. Also, read_only tasks do nothing on perform_setup.

doudou commented on lib/syskit/task_context.rb:
```diff
@@ -672,6 +672,8 @@ def read_current_state
         # Returns true if this component needs to be setup by calling the
         # #setup method, or if it can be used as-is
         def ready_for_setup?(state = nil)
+            return running? if read_only?
```
This is wrong. `running?` refers to the Roby task. When we check if we should set it up, `running?` will always be false.

Look later in this very method. There is a check for the component state to be an exception state, STOPPED or PRE_OPERATIONAL. You would need to check whether the state is a running state in the case of read-only

And add tests ...

jhonasiv committed:
fix: ready for setup if orocos task is in runtime state when read_only

jhonasiv committed:
fix(test): do not configure read only tasks

jhonasiv committed:
feat(test): add tests for read only tasks on #ready_for_setup and #perform_setup

doudou commented on lib/syskit/test/network_manipulation.rb:
```diff
@@ -545,6 +545,8 @@ def syskit_configure(
                     current_pending = pending.size
                     has_missing_states = false
                     pending.delete_if do |t|
+                        next true if t.read_only?
+
```
This *must* not be needed.

doudou commented on test/test_task_context.rb:
```diff
@@ -700,6 +700,34 @@ def push_task_state(state)
+            it "does nothing when the task is read_only" do
+                task.should_receive(:read_only?).and_return(true)
+                task.should_receive(:prepare_for_setup).never
+                flexmock(task.orocos_task).should_receive(:configure).never
```
You still need to actually do something ...

doudou committed:
fix: update read-only tests now that it needs the component to be started before configuration

The change we just made was to require the component to be started
for the read-only task to let itself be configured.

jhonasiv committed:
fix(test): try to start read only task on perform_setup test

jhonasiv committed:
fix(test): check if read_only tasks properly ignore perform_setup

doudou commented on test/test_task_context.rb:
```diff
@@ -2336,6 +2333,16 @@ def assert_process_events_does_not_block
 
+            it "does not perform setup on configuration" do
+                task = deployment.task("test")
+                flexmock(task).should_receive(:perform_setup).once
```
Mocking a method makes it not being called *at all*. I.e. this mock ensures that the next one is never called.

You have to add `.pass_thru` to let the call go through.

jhonasiv committed:
fix(test): add pass thru to perform setup mock

jhonasiv commented:
Ping @doudou
